### PR TITLE
Indexer ES document id

### PR DESCRIPTION
**Pull request by Mikwiss**
Hi @jnioche,

I saw the https://github.com/DigitalPebble/storm-crawler/issues/1025 issue and your answer related to https://github.com/DigitalPebble/storm-crawler/issues/671. That's was exactly the purpose of this PR !

In our case we use a random UUID to store the content (to keep all data). We add a new boolean parameter to handle this : `es.indexer.id.random`. The `getDocumentId()` can also be overrided to create any ID with metadata.

Signed-off-by: Mathieu Bret <mikwiss00@gmail.com>

Thanks for contributing to StormCrawler, your efforts are appreciated!

Developer Certificate of Origin
===============================

By contributing to StormCrawler, you accept and agree to the following terms and conditions (the *Developer Certificate of Origin*) for your present and future contributions submitted to StormCrawler. 
Please refer to the *Developer Certificate of Origin* section in [`CONTRIBUTING.md`](CONTRIBUTING.md) for details.

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

Before opening a PR, please check that: 

* You've squashed your commits into a single one
* You've described what the PR does or at least point to a related issue
* You've signed-ff your commits with 'git commit -s'
* The code is properly formatted with 'mvn git-code-format:format-code -Dgcf.globPattern=**/*'

Thanks!

### CODE REVIEW

**Comment by rzo1** [comment]
I am wondering, if it would be enough to provide

 ```
 protected String getDocumentID(Metadata metadata, String normalisedUrl) {
 //Return current default
}
```

so user's can override as needed without introducing a new config option?

**Comment by Mikwiss** [comment]
It's enough indeed ! Maybe I am the only one to want keep data :)

**Comment by jnioche** [comment]
Yes that would work as well

**Review by rzo1 — APPROVED** [review]

**Comment by jnioche** [comment]
We'll need this for the other modules (SOLR, OpenSearch etc...) so I'd rather we did it for all of them in one go.
The new method could go in _AbstractIndexerBolt_ with the default implementation there, then call it in each implementation.

**Comment by rzo1** [comment]
> We'll need this for the other modules (SOLR, OpenSearch etc...) so I'd rather we did it for all of them in one go.
> The new method could go in _AbstractIndexerBolt_ with the default implementation there, then call it in each implementation.

+1

**Comment by Mikwiss** [comment]
Hi guys, 

I made the change, but I'm not sure about _solr_ and _sql_ (despite changes don't change the behavior).

File: external/solr/src/main/java/com/digitalpebble/stormcrawler/solr/bolt/IndexerBolt.java [hunk 1]
```diff
@@ -104,7 +104,7 @@ public void execute(Tuple tuple) {
                 // Distinguish the value used for indexing
                 // from the one used for the status
                 String normalisedurl = valueForURL(tuple);
-                doc.addField(fieldNameForURL(), normalisedurl);
+                doc.addField(fieldNameForURL(), getDocumentID(metadata, normalisedurl));
```
Comment (jnioche) [inline external/solr/src/main/java/com/digitalpebble/stormcrawler/solr/bolt/IndexerBolt.java:107]: This is to store the URL, not the doc ID. I don't think we should call getDocumentID here, even if it returns the normalised URL.

File: external/sql/src/main/java/com/digitalpebble/stormcrawler/sql/IndexerBolt.java [hunk 2]
```diff
@@ -139,7 +139,7 @@ public void execute(Tuple tuple) {
 
             // send URL as field?
             if (fieldNameForURL() != null) {
-                preparedStmt.setString(1, normalisedurl);
+                preparedStmt.setString(1, getDocumentID(metadata, normalisedurl));
```
Comment (jnioche) [inline external/sql/src/main/java/com/digitalpebble/stormcrawler/sql/IndexerBolt.java:142]: same as above, this is about the URL, not the docID

**Review by jnioche — CHANGES_REQUESTED** [review]

**Comment by jnioche** [comment]
thanks @Mikwiss!

**Pull request merged**